File: problems/560. Subarray Sum Equals K/solution.py
from collections import defaultdict
from typing import List

class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        prefix_sum = 0
        prefix_sum_count = defaultdict(int)
        prefix_sum_count[prefix_sum] += 1
        ans = 0
        for n in nums:
            prefix_sum += n
            if prefix_sum - k in prefix_sum_count:
                ans += prefix_sum_count[prefix_sum-k]
            prefix_sum_count[prefix_sum] += 1
        return ans

File: problems/560. Subarray Sum Equals K/solution.cpp
int subarraySum(vector<int>& nums, int k) {
  int ans = 0;
  int curSum = 0; 
  unordered_map<int, int> map;      
  map[0] = 1;

  for (int num : nums) {
    curSum += num;
    if (map[curSum - k] > 0)
      ans += map[curSum - k];
            
      map[curSum]++;    
  } 

  return ans;
}
